xdsl/dialects/builtin.py

```diff
@@ -513,7 +528,7 @@ def parse_with_type(
         return IntegerAttr(parser.parse_integer(allow_boolean=(type == i1)), type)
 
     def print_without_type(self, printer: Printer):
-        return printer.print(self.value.data)
+        self.type.print_value_without_type(self.value.data, printer)
```

Comment:
The polymorphism feels unnecessary to me

Reply:
What do you mean?

Reply:
I considered it and then felt like doing Python runtime polymorphism is cleaner, if only because we have types like DenseArrayBase, and DenseIntOrFPAttr that will soon have a type and then lots of data, and it seems a little simpler to have the functionality available on the type directly.